File: app/src/main/java/com/manimarank/spell4wiki/activities/MainActivity.java
package com.manimarank.spell4wiki.activities;

import android.app.AlertDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.os.Handler;
import android.view.View;
import android.widget.TextView;

import androidx.appcompat.widget.SearchView;
import androidx.cardview.widget.CardView;

import com.manimarank.spell4wiki.R;
import com.manimarank.spell4wiki.ui.common.BaseActivity;
import com.manimarank.spell4wiki.ui.about.AboutActivity;
import com.manimarank.spell4wiki.utils.pref.AppPref;
import com.manimarank.spell4wiki.utils.GeneralUtils;
import com.manimarank.spell4wiki.utils.NetworkUtils;
import com.manimarank.spell4wiki.utils.pref.PrefManager;
import com.manimarank.spell4wiki.utils.SnackBarUtils;
import com.manimarank.spell4wiki.utils.constants.AppConstants;
import com.manimarank.spell4wiki.utils.constants.Urls;
import com.manimarank.spell4wiki.ui.dialogs.AppLanguageDialog;
import com.manimarank.spell4wiki.ui.dialogs.RateAppDialog;
import com.manimarank.spell4wiki.ui.dialogs.UpdateAppDialog;

public class MainActivity extends BaseActivity implements View.OnClickListener {

    // Views
    SearchView searchView;
    IntentFilter filter = new IntentFilter(AppLanguageDialog.LANGUAGE_FILTER);
    BroadcastReceiver languageChangeReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (!isDestroyed() && !isFinishing() && intent != null && intent.getExtras() != null) {
                String value = intent.getExtras().getString(AppLanguageDialog.SELECTED_LANGUAGE, "");
                if (value != null) {
                    recreate();
                }
            }
        }
    };
    private boolean doubleBackToExitPressedOnce = false;
    private PrefManager pref;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        pref = new PrefManager(MainActivity.this);
        setContentView(R.layout.activity_main);

        initViews();
        registerReceiver(languageChangeReceiver, filter);

        searchView.setQueryHint(getString(R.string.wiktionary_search));
        searchView.setIconifiedByDefault(false);
        searchView.clearFocus();
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                Intent intent = new Intent(getApplicationContext(), WiktionarySearchActivity.class);
                intent.putExtra(AppConstants.SEARCH_TEXT, query);
                startActivity(intent);
                new Handler().postDelayed(() -> {
                    if (searchView != null)
                        searchView.setQuery("", false);
                }, 100);

                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });

        new Handler().post(() -> GeneralUtils.hideKeyboard(MainActivity.this));

        // Update and Rate the app
        if (AppPref.INSTANCE.checkAppUpdateAvailable(MainActivity.this))
            UpdateAppDialog.INSTANCE.show(MainActivity.this);
        else
            RateAppDialog.INSTANCE.show(MainActivity.this);

    }

    @Override
    public void onClick(View view) {
        if (!NetworkUtils.INSTANCE.isConnected(getApplicationContext())) {
            SnackBarUtils.INSTANCE.showLong(view, getString(R.string.check_internet));
            return;
        }

        if (pref.isAnonymous()) {
            SnackBarUtils.INSTANCE.showLong(view, getString(R.string.login_to_contribute));
            return;
        }
        switch (view.getId()) {
            case R.id.card_spell4wiki:
                Intent intentWiki = new Intent(getApplicationContext(), Spell4Wiktionary.class);
                intentWiki.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
                startActivity(intentWiki);
                break;
            case R.id.card_spell4wordlist:
                Intent intentWordList = new Intent(getApplicationContext(), Spell4WordListActivity.class);
                intentWordList.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
                startActivity(intentWordList);
                break;
            case R.id.card_spell4word:
                Intent intentWord = new Intent(getApplicationContext(), Spell4WordActivity.class);
                intentWord.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
                startActivity(intentWord);
                break;
        }
    }

    /**
     * Init views
     */
    private void initViews() {

        CardView cardView1 = findViewById(R.id.card_spell4wiki);
        CardView cardView2 = findViewById(R.id.card_spell4wordlist);
        CardView cardView3 = findViewById(R.id.card_spell4word);
        searchView = findViewById(R.id.search_view);

        cardView1.setOnClickListener(this);
        cardView2.setOnClickListener(this);
        cardView3.setOnClickListener(this);

        TextView txtWelcomeUser = findViewById(R.id.txt_welcome_user);
        txtWelcomeUser.setText(String.format(getString(R.string.welcome_user), pref.getName()));

        findViewById(R.id.btn_about).setOnClickListener(v -> startActivity(new Intent(getApplicationContext(), AboutActivity.class)));
        findViewById(R.id.btn_settings).setOnClickListener(v -> startActivity(new Intent(getApplicationContext(), SettingsActivity.class)));
        findViewById(R.id.btn_logout).setOnClickListener(v -> logoutUser());

        String urlMyContribution = String.format(Urls.COMMONS_CONTRIBUTION, pref.getName());
        TextView btnMyContributions = findViewById(R.id.txtViewMyContribution);
        TextView btnLogin = findViewById(R.id.txtLogin);

        btnMyContributions.setOnClickListener(v -> {
            if (NetworkUtils.INSTANCE.isConnected(getApplicationContext()))
                GeneralUtils.openUrl(MainActivity.this, urlMyContribution, getString(R.string.view_my_contribution));
            else
                SnackBarUtils.INSTANCE.showNormal(btnMyContributions, getString(R.string.check_internet));
        });
        btnLogin.setOnClickListener(v -> {
            if (NetworkUtils.INSTANCE.isConnected(getApplicationContext()))
                pref.logoutUser();
            else
                SnackBarUtils.INSTANCE.showNormal(searchView, getString(R.string.check_internet));
        });

        View viewContribute = findViewById(R.id.layoutContributeOptions);
        View viewLogin = findViewById(R.id.layoutLogin);
        if (pref.isAnonymous()) {
            //viewContribute.setVisibility(View.GONE);
            btnMyContributions.setVisibility(View.GONE);
            txtWelcomeUser.setVisibility(View.GONE);
            findViewById(R.id.btn_logout).setVisibility(View.GONE);
            viewLogin.setVisibility(View.VISIBLE);
        }
    }

    private void logoutUser() {
        if (NetworkUtils.INSTANCE.isConnected(getApplicationContext())) {
            new AlertDialog.Builder(this)
                    .setTitle(R.string.logout_confirmation)
                    .setMessage(R.string.logout_message)
                    .setPositiveButton(getString(R.string.yes), (dialog, which) -> {
                        // Logout user
                        logoutApi();
                        pref.logoutUser();
                    })
                    .setNegativeButton(R.string.no, null)
                    .show();
        } else
            SnackBarUtils.INSTANCE.showNormal(searchView, getString(R.string.check_internet));
    }

    private void logoutApi() {
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (searchView != null) {
            searchView.clearFocus();
            GeneralUtils.hideKeyboard(this);
        }
        if (!NetworkUtils.INSTANCE.isConnected(getApplicationContext()))
            SnackBarUtils.INSTANCE.showNormal(searchView, getString(R.string.check_internet));
    }

    @Override
    public void onBackPressed() {
        if (doubleBackToExitPressedOnce)
            super.onBackPressed();
        else {
            this.doubleBackToExitPressedOnce = true;
            SnackBarUtils.INSTANCE.showLong(searchView, getString(R.string.alert_to_exit));
        }
        new Handler().postDelayed(() -> doubleBackToExitPressedOnce = false, 2000);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (languageChangeReceiver != null)
            unregisterReceiver(languageChangeReceiver);
    }
}

File: app/src/main/java/com/manimarank/spell4wiki/ui/main/MainActivity.kt
package com.manimarank.spell4wiki.ui.main

import android.app.AlertDialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.os.Handler
import android.view.View
import androidx.appcompat.widget.SearchView
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import com.manimarank.spell4wiki.R
import com.manimarank.spell4wiki.data.prefs.AppPref.INSTANCE.checkAppUpdateAvailable
import com.manimarank.spell4wiki.data.prefs.PrefManager
import com.manimarank.spell4wiki.ui.about.AboutActivity
import com.manimarank.spell4wiki.ui.common.BaseActivity
import com.manimarank.spell4wiki.ui.dialogs.AppLanguageDialog
import com.manimarank.spell4wiki.ui.dialogs.RateAppDialog
import com.manimarank.spell4wiki.ui.dialogs.UpdateAppDialog
import com.manimarank.spell4wiki.ui.settings.SettingsActivity
import com.manimarank.spell4wiki.ui.spell4wiktionary.Spell4Wiktionary
import com.manimarank.spell4wiki.ui.spell4wiktionary.Spell4WordActivity
import com.manimarank.spell4wiki.ui.spell4wiktionary.Spell4WordListActivity
import com.manimarank.spell4wiki.ui.spell4wiktionary.WiktionarySearchActivity
import com.manimarank.spell4wiki.utils.EdgeToEdgeUtils.setupStatusBarHandling
import com.manimarank.spell4wiki.utils.GeneralUtils.hideKeyboard
import com.manimarank.spell4wiki.utils.GeneralUtils.openUrl
import com.manimarank.spell4wiki.utils.NetworkUtils.isConnected
import com.manimarank.spell4wiki.utils.SnackBarUtils.showLong
import com.manimarank.spell4wiki.utils.SnackBarUtils.showNormal
import com.manimarank.spell4wiki.utils.constants.AppConstants
import com.manimarank.spell4wiki.utils.constants.Urls
import com.manimarank.spell4wiki.utils.makeGone
import com.manimarank.spell4wiki.utils.makeVisible
import com.manimarank.spell4wiki.databinding.ActivityMainBinding

class MainActivity : BaseActivity(), View.OnClickListener {

    private lateinit var binding: ActivityMainBinding
    // Views
    private var languageChangeReceiver: BroadcastReceiver? = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            if (!isDestroyed && !isFinishing && intent.extras != null) {
                val value = intent.extras?.getString(AppLanguageDialog.SELECTED_LANGUAGE, "")
                if (value != null) {
                    recreate()
                }
            }
        }
    }
    private var doubleBackToExitPressedOnce = false
    private lateinit var pref: PrefManager
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        pref = PrefManager(this@MainActivity)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Setup proper status bar handling
        setupStatusBarHandling(binding.root)

        initViews()
        languageChangeReceiver?.let {
            LocalBroadcastManager.getInstance(this).registerReceiver(
                it, IntentFilter(AppLanguageDialog.SELECTED_LANGUAGE))
        }
        binding.searchView.queryHint = getString(R.string.wiktionary_search)
        binding.searchView.setIconifiedByDefault(false)
        binding.searchView.clearFocus()
        binding.searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                val intent = Intent(applicationContext, WiktionarySearchActivity::class.java)
                intent.putExtra(AppConstants.SEARCH_TEXT, query)
                startActivity(intent)
                Handler().postDelayed({ binding.searchView.setQuery("", false) }, 100)
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                return false
            }
        })
        Handler().post { hideKeyboard(this@MainActivity) }

        // Update and Rate the app
        if (checkAppUpdateAvailable(this@MainActivity)) UpdateAppDialog.show(this@MainActivity) else RateAppDialog.show(this@MainActivity)
    }

    override fun onClick(view: View) {
        if (!isConnected(applicationContext)) {
            showLong(view, getString(R.string.check_internet))
            return
        }
        if (pref.isAnonymous == true) {
            showLong(view, getString(R.string.login_to_contribute))
            return
        }
        when (view.id) {
            R.id.card_spell4wiki -> {
                val intentWiki = Intent(applicationContext, Spell4Wiktionary::class.java)
                intentWiki.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
                startActivity(intentWiki)
            }
            R.id.card_spell4wordlist -> {
                val intentWordList = Intent(applicationContext, Spell4WordListActivity::class.java)
                intentWordList.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
                startActivity(intentWordList)
            }
            R.id.card_spell4word -> {
                val intentWord = Intent(applicationContext, Spell4WordActivity::class.java)
                intentWord.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
                startActivity(intentWord)
            }
        }
    }

    /**
     * Init views
     */
    private fun initViews() {
        binding.cardSpell4wiki.setOnClickListener(this)
        binding.cardSpell4wordlist.setOnClickListener(this)
        binding.cardSpell4word.setOnClickListener(this)
        binding.txtWelcomeUser.text = String.format(getString(R.string.welcome_user), pref.name)
        binding.btnAbout.setOnClickListener { startActivity(Intent(applicationContext, AboutActivity::class.java)) }
        binding.btnSettings.setOnClickListener { startActivity(Intent(applicationContext, SettingsActivity::class.java)) }
        binding.btnLogout.setOnClickListener { logoutUser() }
        val urlMyContribution = String.format(Urls.COMMONS_CONTRIBUTION, pref.name)
        binding.txtViewMyContribution.setOnClickListener { if (isConnected(applicationContext)) openUrl(this@MainActivity, urlMyContribution, getString(R.string.view_my_contribution)) else showNormal(binding.txtViewMyContribution, getString(R.string.check_internet)) }
        binding.txtLogin.setOnClickListener { if (isConnected(applicationContext)) pref.logoutUser() else showNormal(binding.searchView, getString(R.string.check_internet)) }
        if (pref.isAnonymous == true) {
            binding.txtViewMyContribution.makeGone()
            binding.txtWelcomeUser.makeGone()
            binding.btnLogout.makeGone()
            binding.layoutLogin.makeVisible()
        }
    }

    private fun logoutUser() {
        if (isConnected(applicationContext)) {
            AlertDialog.Builder(this)
                    .setTitle(R.string.logout_confirmation)
                    .setMessage(R.string.logout_message)
                    .setPositiveButton(getString(R.string.yes)) { _: DialogInterface?, _: Int ->
                        // Logout user
                        logoutApi()
                        pref.logoutUser()
                    }
                    .setNegativeButton(R.string.no, null)
                    .show()
        } else showNormal(binding.searchView, getString(R.string.check_internet))
    }

    private fun logoutApi() {}
    override fun onResume() {
        super.onResume()
        binding.searchView.clearFocus()
        hideKeyboard(this)
        if (!isConnected(applicationContext)) showNormal(binding.searchView, getString(R.string.check_internet))
    }

    override fun onBackPressed() {
        if (doubleBackToExitPressedOnce) 
            super.onBackPressed() 
        else {
            doubleBackToExitPressedOnce = true
            showLong(binding.searchView, getString(R.string.alert_to_exit))
        }
        Handler().postDelayed({ doubleBackToExitPressedOnce = false }, 2000)
    }

    override fun onDestroy() {
        super.onDestroy()
        languageChangeReceiver?.let {
            LocalBroadcastManager.getInstance(this).unregisterReceiver(it)
        }
    }
}